File: SmartOTT.cpp
/**
 * @file SmartOTT.cpp
 * @author Sidharth S
 * @version 0.1
 * @date 2022-07-28
 *
 *
As a smart OTT company, we want to offer customized OTT plans for our customers
so that the customers should be able to choose to build their plan by choosing different
streaming platforms and how many viewing hours they want for each of them.
We do support the following streaming services and below is their tariff plan.

Streaming Service | Price per unit    | Unit in hours
Netflix           | Rs. 10            | 10 hrs
Amazon Prime      |   Rs. 2           | 5 hrs
Hotstar           | Rs. 1             | 5 hrs

Viewing hours can be subscribed based on the multiples of the above mentioned units only.
Based on the user’s choice, the system should show the total amount which should be paid by
the customer.

Sample Input 1
Viewing hours for Netflix: 20
Viewing hours for Amazon Prime: 10
Viewing hours for Hotstar: 50
Sample Output
Total amount should be paid: Rs.34 (10 * 2 + 2 * 2 + 1 * 10)

Sample Input 2
Viewing hours for Netflix: 10
Viewing hours for Amazon Prime: 0
Viewing hours for Hotstar: 100
Sample Output
Total amount should be paid: Rs.30 (10 * 1 + 2 * 0 + 1 * 20)

Sample Input 3
Viewing hours for Netflix: 10
Viewing hours for Amazon Prime: 2
Sample Output
Amazon Prime allows viewing hours in multiples of 5 only[3] [4]
 *
 *
 */

#include <iostream>
#include <vector>

using namespace std;

typedef enum
{
    failed,
    success
} SubscriptionStatus;

class Subscription
{
private:
    SubscriptionStatus status;
    int amount;

public:
    SubscriptionStatus getStatus() const
    {
        return this->status;
    }
    int getAmount() const
    {
        return this->amount;
    }
    Subscription(SubscriptionStatus status = failed, int amount = 0)
    {
        this->status = status;
        this->amount = amount;
    }
};

class StreamingService
{
private:
    string name;
    int unitInHours;
    int pricePerUnit;
    int noOfUnits(int viewingHours) const
    {
        if (unitInHours != 0)
            return (int)viewingHours / unitInHours;
        return 0;
    }
    bool validate_viewing_hours(int viewingHours) const
    {
        if (unitInHours != 0)
            return viewingHours % unitInHours == 0;
        return false;
    }
    int calculate_amount(int viewingHours) const
    {
        return unitInHours * noOfUnits(viewingHours);
    }

public:
    string getName() const
    {
        return name;
    }
    int getUnitInHours() const
    {
        return unitInHours;
    }
    Subscription subscribe(int viewingHours)
    {
        int amount = 0;
        SubscriptionStatus status = failed;
        bool viewingHoursValid = validate_viewing_hours(viewingHours);
        if (viewingHoursValid)
        {
            amount = calculate_amount(viewingHours);
            status = success;
        }
        return Subscription(status, amount);
    }
    StreamingService(string name, int pricePerUnit, int unitInHours)
    {
        this->name = name;
        this->pricePerUnit = pricePerUnit;
        this->unitInHours = unitInHours;
    }
};

class SmartOtt
{
private:
    vector<StreamingService> streamingServices;

public:
    Subscription build_subscription()
    {
        int hours, totalAmount = 0;
        for (StreamingService service : streamingServices)
        {
            cout << "Viewing hours for " << service.getName() << " : ";
            cin >> hours;
            Subscription serviceSub = service.subscribe(hours);
            if (serviceSub.getStatus() == failed)
            {
                cout << service.getName() << " allows viewing hours in multiples of " << service.getUnitInHours() << " only!" << endl;
                return Subscription(failed, 0);
            }
            totalAmount += serviceSub.getAmount();
        }
        return Subscription(success, totalAmount);
    }
    SmartOtt(vector<StreamingService> streamingServices)
    {
        this->streamingServices = streamingServices;
    }
};

int main(int argc, char *argv[])
{
    StreamingService netflix = StreamingService("Netflix", 10, 10);
    StreamingService amazonPrime = StreamingService("Amazon Prime", 2, 5);
    StreamingService hotstar = StreamingService("Hotstar", 1, 5);

    vector<StreamingService> streamingServices = {netflix, amazonPrime, hotstar};

    SmartOtt smartOtt = SmartOtt(streamingServices);
    Subscription new_subscription = (Subscription)smartOtt.build_subscription();
    if (new_subscription.getStatus() == success)
        cout << "Rs." << new_subscription.getAmount();
    return 0;
}

// class SubscriptionWithSplitUp : public Subscription
// {
//     private:
//     string splitUp;
//     public:
//     SubscriptionWithSplitUp()
//     :Subscription()
//     {

//     }
// }

File: SmartOTT.py
# Smart OTT offering custom streaming platforms

# class SmartOTTp:
#     def __init__(self,):
#         self.viewing_hours_for_netflix = 0
#         self.viewing_hours_for_amazon_prime = 0
#         self.viewing_hours_for_hotstar = 0
#         self.split = ""
#         self.total_cost = 0

#     def build(self):
#         self.viewing_hours_for_netflix = int(
#             input("Viewing hours for netflix: "))
#         if self.viewing_hours_for_netflix % 10 != 0:
#             return "Viewing hours for netflix should be in multiples of 10"
#         self.viewing_hours_for_amazon_prime = int(
#             input("Viewing hours for amazon prime: "))
#         if self.viewing_hours_for_amazon_prime % 5 != 0:
#             return "Viewing hours for amazon prime should in multiples of 5"
#         self.viewing_hours_for_hotstar = int(
#             input("Viewing hours for hotstar: "))
#         if self.viewing_hours_for_hotstar % 5 != 0:
#             return "Viewing hours for hotstar should be in multiples of 5"

#         netflix_units = self.viewing_hours_for_netflix // 10
#         amazon_prime_units = self.viewing_hours_for_amazon_prime // 5
#         hotstar_units = self.viewing_hours_for_hotstar // 5

#         self.split = "( 10 * " + str(netflix_units) + " + 5 * " + \
#             str(amazon_prime_units) + " + 5 * " + str(hotstar_units) + " )"
#         self.total_cost = netflix_units * 10 + \
#             amazon_prime_units * 5 + hotstar_units * 5
#         return "Total amount should be paid: Rs." + str(self.total_cost) + " " + self.split


# smart_ott = SmartOTT()
# plan = smart_ott.build()
# print(plan)

from typing import List, Tuple


class StreamingService:
    def __init__(self, name, unitInHours, pricePerUnit):
        self._unitInHours = unitInHours
        self._pricePerUnit = pricePerUnit
        self._name = name
        self._viewingHours = 0
        self._amount = 0

    @property
    def name(self) -> str:
        return self._name

    @property
    def unitInHours(self) -> int:
        return self._unitInHours

    @property
    def pricePerUnit(self) -> int:
        return self._unitInHours

    @property
    def viewingHours(self) -> int:
        return self._viewingHours

    @property
    def noOfUnits(self) -> int:
        return 0 if self.viewingHours == 0 else (self.viewingHours // self.unitInHours)

    @property
    def amount(self) -> int:
        return self._amount

    def _validate_viewing_hours(self, hours) -> bool:
        if hours % self.unitInHours == 0:
            self._viewingHours = hours
            return True
        return False

    def _calculate_amount(self, hours) -> str:
        if self._validate_viewing_hours(int(hours)):
            self._amount = self.noOfUnits * self.pricePerUnit
            split_up = str(self.noOfUnits) + " * " + str(self.pricePerUnit)
            return (self.amount, split_up)
        else:
            raise ValueError("%s allows viewing hours in multiples of %d only" % (
                self.name, self.unitInHours))

    def subscribe(self, hours) -> Tuple[int, str]:
        return self._calculate_amount(hours)


class SmartOTT:
    def __init__(self, streamingServices: List[StreamingService]):
        self._streamingServices = [*streamingServices]
        self._total_amount = 0
        self._split_up = []

    @property
    def totalAmount(self) -> str:
        return "Rs." + str(self._total_amount)

    @property
    def splitUp(self) -> str:
        return ' + '.join(self._split_up)

    @property
    def bill(self) -> str:
        return self.totalAmount + " " + self.splitUp

    def build_subscription(self) -> Tuple[str, str]:
        for service in self._streamingServices:
            hours = int(input("Viewing hours for %s : " % service.name))
            try:
                amount, split_up = service.subscribe(hours)
                self._total_amount += amount
                self._split_up.append(split_up)
            except ValueError as e:
                print(e)
                return ("failed", "")
        return ("success", self.bill)


netflix = StreamingService("Netflix", 10, 10)
amazonPrime = StreamingService("Amazon Prime", 5, 2)
hotstar = StreamingService("Hotstar", 5, 1)

smart_ott = SmartOTT([netflix, amazonPrime, hotstar])
subscription_status, bill = smart_ott.build_subscription()
if subscription_status == "success":
    print(bill)
